### ISSUE TITLE
Add `display-notation`

### ISSUE THREAD

**Issue by romainmenke** [stylelint/stylelint#8958]
### What is the problem you're trying to solve?

The `display` property has several legacy values:

https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Values/display-legacy

```
<display-legacy> = 
  inline-block  |
  inline-table  |
  inline-flex   |
  inline-grid 
```

The modern equivalent is to use space separation:

```css
.foo {
  display: inline flex;
}
```

With the upcoming grid lanes (a.k.a. masonry) feature some in the CSSWG are concerned that authors will not intuit that `display: inline-grid-lanes` isn't valid and that they should write `display: inline grid-lanes` instead.

### What solution would you like to see?

A new rule to enforce consistency in notation for values in the `display` property.

This rule could have a primary rule option for either `modern` or `legacy`. With `modern` being the default.

When set to `modern` these legacy notations are errors:

```css
.foo {
  display: inline-block;
  display: inline-table;
  display: inline-flex;
  display: inline-grid;
}
```

And possibly also these combinations which never existed:

```css
.foo {
  display: inline-grid-lanes;
  display: inline-flow-root;
}
```

Adding this last list can help authors better understand the new syntax through error messages and auto fixes.

When set to `legacy` these mondern notations are errors:

```css
.foo {
  display: inline flow-root;
  display: inline table;
  display: inline flex;
  display: inline grid;
}
```

But not these:

```css
.foo {
  display: inline grid-lanes;
}
```

**Comment by jeddy3** [comment]
Sounds good.

> And possibly also these combinations which never existed

Let's limit the rule to known values for now. Converting those will be a good first step towards making people more familiar with the ~~two-value~~ modern syntax.

We can always revisit this decision if it emerges that `declaration-property-value-no-unknown` catching the likes of `display: inline-grid-lanes;` isn't useful enough.

Rule details:

- Name: `property-display-notation`
- Primary option: `"modern"|"legacy"`
- Secondary options: None
- Autofixable: Yes
- Message: `Expected "${unfixed}" to be "${fixed}"`
- Description: "Specify modern or legacy notation for the display property"
- Extended description: Something about the introduction of the ~~two-value~~ modern syntax and how it'll used going forward
- Section: "Enforce conventions" -> "Notation"
- Config: Yes, in standard with `"modern"` ([Baseline ref](https://caniuse.com/wf-two-value-display))

**Comment by firefoxic** [comment]
> the two-value syntax

Just to be clear, it would be more correct to call it not _two-value syntax_, or even _multi-value syntax_ or _multi-keyword syntax_, but simply _modern syntax_, because in some cases there can be three keywords in the value:
```css
.foo {
	display: inline list-item flow-root;
}
.bar {
	display: block list-item flow;
}
```
(and this is valid), but some values can only have one keyword:
```css
.foo {
	display: none;
}
.bar {
	display: contents;
}
```
Such cases also need to be handled in the proposed rule.

**Comment by romainmenke** [comment]
Hi @firefoxic, 

I am unsure what your point is :)

I think the most succinct way to describe the proposed rule is to either enforce or disallow the use of the legacy display keywords:

```
<display-legacy> = 
  inline-block  |
  inline-table  |
  inline-flex   |
  inline-grid 
```

The proposed rule name is `property-display-notation` with the options of `modern` or `legacy`.

We don't plan to use `two-value syntax` as a description or term in this rule, but the `two-value` term stems from `inline-block` vs. `inline block`.

It doesn't imply that the `display` property is limited to only two keywords. Only that `inline-block` can be written with two values instead of one.

**Comment by jeddy3** [comment]
My bad, I confused things by including "two-syntax" in [my comment](https://github.com/stylelint/stylelint/issues/8958#issuecomment-3729772564) 😅 . I've deleted it in favour of "modern".

**Comment by firefoxic** [comment]
> I am unsure what your point is :)
> 
> I think the most succinct way to describe the proposed rule is to either enforce or disallow the use of the legacy display keywords

In general, I am very much looking forward to this rule, and I just wanted to highlight such cases (which I have cited) that are not obvious at first glance.

But now I am wondering: does `modern` simply mean prohibiting the few values you have cited with the prefix `inline-`? If so, that is not enough for me. I would then like an additional option, `strict` or `verbose`, which would prohibit short forms such as `block`, `grid`, `flex`, `flow-root`, `list-item`, etc., and would require (with autofix) `block flow`, `block grid`, `block flex`, `block flow-root`, `block list-item flow`, etc. instead.

I've been planning to write (and offer you) such a rule for two years now, but I haven't had the time.

By the way, lightningcss perfectly converts modern syntax to legacy for production.

**Comment by romainmenke** [comment]
My immediate concern was `inline-block` vs. `inline block`.
This is in response to the CSSWG considering adding `inline-grid-lanes`.
Most of the CSSWG (myself included) do not want this.

Some members are concerned that authors haven't adopted `inline block` yet and will try `inline-grid-lanes`, only to find that it doesn't work. They might then think that `grid-lines` can't be used `inline` and not even try `inline grid-lanes`.

Teaching authors that they can write `inline block` and `inline grid` will help the CSSWG avoid introducing more new keywords that are ironically also immediately legacy.

I think this is separate from the more verbose forms like `block flow` although I do see the value in this :)

I am unsure if this should be a separate rule or if maybe my proposed rule should instead be scoped to the value instead of the `display` property?

**Comment by firefoxic** [comment]
> Most of the CSSWG (myself included) do not want this.

Totally on your side!

> Teaching authors that they can write `inline block` and `inline grid` will help the CSSWG avoid introducing more new keywords that are ironically also immediately legacy.
> 
> I think this is separate from the more verbose forms like `block flow` although I do see the value in this :)

I see educational value in strictly requiring complete modern notation for the `display` property.

My experience has shown that simply prohibiting these four values

```json
{
	"rules": {
		"declaration-property-value-disallowed-list": {
			"display": [
				"inline-block",
				"inline-table",
				"inline-flex",
				"inline-grid"
			]
		}
	}
}
```

did not give my colleagues any understanding of why they were prohibited, let alone how the `display` property actually works. So I moved on to a more radical prohibition:

```json
{
	"rules": {
		"declaration-property-value-disallowed-list": {
			"display": [
				"inline-block",
				"inline-table",
				"inline-flex",
				"inline-grid",
				"inline",
				"block",
				"flex",
				"grid",
				"table",
				"list-item",
				"flow-root",
			]
		}
	}
}
```

This worked better because there were more questions and more answers from me. But what is still missing is autofix to the complete modern syntax without omitting any parts of the value. After all, if the author writes (out of old habit or ignorance, it doesn't matter) `display: block;`, and the linter automatically adds `display: block flow;`, then this teaches the device the property both at the time of writing and when reading the code base (automatically corrected when the rule is implemented).

By the way, in this regard, I would like to have one more additional option in the rule — a requirement for the order of keywords. There are different options here, but I like the order `<display-outside> <display-inside>` the most, and if there is `list-item`, then it is placed between them. This will add consistency and further contribute to the educational aspect.

**Comment by Loirooriol** [comment]
> `inline-block` vs. `inline block`

There is no `inline block`, because both `inline` and `block` are outer display values. The modern syntax is `inline flow-root`.

> if there is `list-item`, then it is placed between them

I think that's weird, better place it at the end like in [the grammar](https://drafts.csswg.org/css-display-3/#typedef-display-listitem):

```
<display-listitem> = <display-outside>? && [ flow | flow-root ]? && list-item
```

This way it's clear that `block flow` and `block flow list-item` generate the same kind of box, it's just that the latter is a list item.

> So I moved on to a more radical prohibition

I think forbidding `block` but not `inline` is inconsistent.

I think that forbidding `list-item` and `block` but not `block list-item` is inconsistent.

**Comment by romainmenke** [comment]
Thank you for chiming in here @Loirooriol

> There is no `inline block`, because both `inline` and `block` are outer display values. The modern syntax is `inline flow-root`.

I misread this: https://drafts.csswg.org/css-display-3/#the-display-properties:~:text=level%20block%20container-,aka%20inline%20block,-run%2Din

> `inline-level` `block container` aka `inline block`

**Comment by firefoxic** [comment]
> There is no `inline block`

Yes, and the rule should prohibit this (without auto-fix, because it is unclear what the author meant). Or is this already a matter of the rule about non-existent values for properties?

> This way it's clear that `block flow` and `block flow list-item` generate the same box, it's just that the latter is a list item.

Yours, like any other option, has a right to exist. Neither the specification nor browsers restrict us in any way as to the order in which keywords should be listed. But consistency in code is always important.

> I think forbidding `block` but not `inline` is inconsistent.

Hmm... The separate keyword `inline` is also banned for me. Or did I misunderstand?

> I think that forbidding `list-item` and `block` but not `block list-item` is inconsistent.

I agree with that, as well as with the absence of `list-item flow-root`, for example. But it's difficult to take all cases into account in that rule, and my colleagues aren't happy with such a list either. That's why we need a separate rule with autofix and possibly an error for obviously incorrect or confusing values.

**Comment by Loirooriol** [comment]
> Yours, like any other option, has a right to exist

Of course, you do you. I just think it's weird to place `list-item` at the middle, so I don't want the lint to recommend that.

> The separate keyword `inline` is also banned for me

Sorry, I somehow missed it.

**Comment by romainmenke** [comment]
So the proposed options are:

- `legacy`: use single keywords like `inline-block`
- `modern`: use multi-keyword values like `inline flow-root` but otherwise use as few keywords as needed by omitting defaults
- `full/verbose`: use multi-keyword syntax and do not omit defaults

| legacy | modern | full/verbose |
| --- | --- | --- |
| `inline-block` | `inline flow-root` | `inline flow-root` |
| `block` | `block` | `block flow` |
| `list-item` | `list-item` | `block flow list-item` |

@jeddy3 @ybiquitous Do we have precedence of a rule with three or more options?
Are there concerns around adding the `full/verbose` option?

-----

>  I just think it's weird to place `list-item` at the middle, so I don't want the lint to recommend that.

I agree
Using the order as defined in the syntax definitions makes most sense to me.

Any autofix should follow that order.
But anyone can also add/create a separate linter/formatter rule to enforce a different order.

The autofix considered here wouldn't re-order values that don't have issues.
So it might fix `list-item` into `block flow list-item`. But it won't re-order `block list-item flow`.

**Comment by firefoxic** [comment]
> Any autofix should follow that order. But anyone can also add/create a separate linter/formatter rule to enforce a different order.
> 
> The autofix considered here wouldn't re-order values that don't have issues. So it might fix `list-item` into `block flow list-item`. But it won't re-order `block list-item flow`.

I agree that sorting keywords is more a matter of stylistic preference.

**Comment by ybiquitous** [comment]
I think providing the two options (`legacy`/`modern`) only is better for UX.

- Can we consider `full/verbose` as a kind of `modern`?
- Is `full/verbose` worth it for many folks or a few experts?

**Comment by firefoxic** [comment]
> Can we consider `full/verbose` as a kind of `modern`?

Before this issue, verbose was the modern syntax for me. And I see the logic in that. Because omitting `block` before `grid` is not at all obvious from a common sense point of view (rather than our habits), since `display` has an initial value of `inline` (which is equivalent to `inline flow`), and not `block` (`block flow`). Why then would `grid` expand to `block grid`? Since `grid` is `<display-inside>`, it should only override the `<display-inside>` part and leave `<display-outside>` untouched. But this is not the case. `display: grid;` does not follow the modern principle of value composition and expands to `display: block grid;` simply for historical reasons. Therefore, `display: grid;` is as much legacy syntax to me as `display: inline-block;`.

>  Is `full/verbose` worth it for many folks or a few experts?

In my opinion, experts can easily work with any syntax, even when different ones are mixed in the same source code. But most people need a complete modern syntax (with autofix, of course) for educational purposes.

So yes, I agree with that:

> providing the two options (`legacy`/`modern`) only is better for UX.

… not just for UX 🙂

**Comment by romainmenke** [comment]
`full/verbose` is more opinionated (I think) than the proposed `modern` option.

I was hoping to make this rule and a config option that disallows `inline-block` part of the standard config and I think many users will be surprised when `block` also becomes `block flow` when using the standard config.

I do agree with @firefoxic that the `full/verbose` option is more interesting because explicit/verbose notations makes all parts visible. It makes it easier to really understand the `display` property.

But most users actually prefer brevity, as we see with shorthands, nesting, ...

------

I would rather have only `legacy`/`modern` than only `legacy`/`verbose`.

But this concern/preference is limited in time.

I think it is important at this point in time that authors learn about `inline flow-root` so that we don't have to keep introducing new legacy keywords like `inline-grid-lanes`.

I am unsure what my preference or opinion would be if authors had already widely adopted `inline flow-root`.

------

Some alternative approaches for having three rule options:

### 1 Have a rule to disallow legacy syntax.

Maybe this could be part of `declaration-property-value-keyword-no-deprecated`?
I am unsure if the distinction between `legacy` and `deprecated` is significant for our purposes.

This would disallow `inline-block` and promote `inline flow-root`.

Another rule can then focus on the `short` vs. `full` syntax for the display property.


### 2 leave either `modern` or `full` up to a community supported rule

We can limit standard stylelint support to a rule that makes most sense for our context and leave the remainder up to a community supported rule.

Slightly defeats the purpose as we all want users to learn about something and if you don't know about it, you won't go looking for a stylelint rule for it :)

**Comment by firefoxic** [comment]
> many users will be surprised when `block` also becomes `block flow`

Yes! Just as they will be surprised when `inline-block` becomes `inline flow-root`. They will be surprised and go looking for the reason. They will read the description of the stylelint rule and the article on MDN.

Actually, values starting with `inline-` are not used very often (according to my observations). And changing `block` to `block flow` and `grid` to `block grid` won't break anything, but it will make people read about it earlier, which means they will learn about `inline flow-root` earlier 😉

**Comment by ybiquitous** [comment]
Thank you for providing more insights, which help me learn the modern syntax and background of `display`.

I understand this request's main purpose is to rewrite `<display-legacy>` to the modern notations, e.g., `inline-block` → `inline flow-root`. Considering this, providing the two options `modern`/`legacy` is reasonable and should help many users learn the modern syntax. Also, adding `modern` to `stylelint-config-standard` makes sense, as [the two-value `display` is in Baseline 2023](https://caniuse.com/wf-two-value-display).

```jsonc
{
  "property-display-notation": "modern" // 😄 simple and clear
}
```

By the way, I don't think this should be supported by `declaration-property-value-keyword-no-deprecated` because `<display-legacy>` has been widely used and will not be obsolete in the CSS spec.

---

On the other hand, the second `full` (or `modern-full`?) option is regarding an opinionated issue, as @romainmenke said. One may prefer `display: block` for simplicity, another may prefer `display: block flow` for comprehensiveness. Also, people's preferences may differ, depending on values, e.g., some may prefer `display: block flow` and `display: list-item` simultaneously, etc. I think one option (e.g., `full`) would be quite challenging to achieve such a preference.

```jsonc
{
  "property-display-notation": "full" // 😕 not clear
}
```

Otherwise, secondary options might help, but more discussions would be needed, for example:

```jsonc
{
  "property-display-notation": ["modern", { "verbose": true }] // 🤔 clear?
}
```

Therefore, I think the first implementation should provide the two options `modern`/`legacy` only without any secondary options. We could enhance the rule later for opinionated syntax patterns that folks might prefer more.

**Comment by jeddy3** [comment]
Lots of good points are being made by everyone.

> Do we have precedence of a rule with three or more options?

Yes ([`font-family-name-quotes`](https://stylelint.io/user-guide/rules/font-family-name-quotes) is one I can think of off hand), but we typically use two primary options and then the [`except` and `ignore` secondary options](https://stylelint.io/developer-guide/rules#keyword-ignore-and-except).

When defining a rule, we've tended to separate its definition from its configuration in our shared configs. Rules are [strict by default and made more permissive through secondary options](https://stylelint.io/developer-guide/rules#write-the-rule). 

We may then choose to make them more permissive in our config, [for example](https://github.com/stylelint/stylelint-config-standard/blob/fbbe7b8e3efaf6ad84084268944be412e4bee3df/index.js#L125-L131):

```js
{
  'value-no-vendor-prefix': [ true, {
    ignoreValues: ['box', 'inline-box'],
  }]
}
```

Or invert the primary option for some patterns, [for example](https://github.com/stylelint/stylelint-config-standard/blob/fbbe7b8e3efaf6ad84084268944be412e4bee3df/index.js#L4-L15
):

```js
{
  'alpha-value-notation': [ 'percentage', { 
    exceptProperties: ['opacity', 'fill-opacity', 'flood-opacity', 'stop-opacity', 'stroke-opacity',] 
  }],
}
```

This approach has allowed us to support the broadest range of use cases (and the strictest) rather than just the most common. The majority of users just extend our shared config, whilst a minority use the rules directly (and seemingly appreciate their default strictness, or at least the consistency of strictness and options).

A rule typically targets a section of a specification and uses the names defined within it. In this case, it could be [the section and table containing](https://drafts.csswg.org/css-display-3/#the-display-properties): `short`, `full` (and `legacy`). 

So, a strict-by-default rule could be: `"display-notation": "short"|"full"`. It enforces one column of the table or the other.

The primary option can then be inverted through a ("user-defined") `exceptValues: []`  secondary option. For example, to generally prefer `full` except for `list-item`:

```js
{
  'display-notation': ['full', { 
    exceptValues: ['list-item'] 
  }],
}
```

If we like, we can add a convenience ("keyword") `except: ["legacy-values"]` secondary option for `inline-block`, `inline-table`, `inline-flex` and `inline-grid`:

```js
{
  'display-notation': ['short', { 
    except: ["legacy-values"] 
  }]
}
```

This can help us highlight to people how the rule can be used to modernise legacy keywords.

It seems we can meet both use cases (and the additional ones @ybiquitous mentioned) by designing for the strictest (the "full" request) and adding a secondary convenience option to disallow legacy values. Although the original request was for the latter, we can broaden the scope of the rule to align with our typical methods and accommodate both.

Whatever approach we go for, let's make the extended description detailed and include links to the specs and any relevant articles.

**Comment by ybiquitous** [comment]
This strict-by-default approach sounds reasonable to me 👍

**Comment by jeddy3** [comment]
As there thumb ups, I'll detail a new blueprint:

- Name: `display-notation`
- Primary option: `"short"|"full"`
- Secondary options: 
   - `except: ["legacy-values"]`
   - `ignore: ["non-legacy-values"]`
   - ~~`exceptValues: []`~~
   - ~~`ignoreValues: []`~~
- Autofixable: Yes
- Message: `Expected "${unfixed}" to be "${fixed}"`
- Description: "Specify short or full notation for display values"
- Extended description: Extensive and including at least links to:
   - https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Display/Multi-keyword_syntax
   - https://drafts.csswg.org/css-display-3/#the-display-properties
- Section: "Enforce conventions" -> "Notation"
- Config: Yes, eventually standard with TBD ([Baseline ref](https://caniuse.com/wf-two-value-display))

I've opened https://github.com/stylelint/stylelint-config-standard/issues/387 to discuss which options we should use in our standard config. We don't need to rush it. I suggest we release the rule first and use the opportunity to direct people to https://github.com/stylelint/stylelint-config-standard/issues/387 using the release summary. Something along the lines of:

> It adds the `display-notation` rule. We'd like to [hear your thoughts](https://github.com/stylelint/stylelint-config-standard/issues/387) on which option we should use for it in our [standard config](https://www.npmjs.com/package/stylelint-config-standard)."

**Comment by romainmenke** [comment]
Thank you for the updated blueprint 🙇 

-----

I suggest that we make `exceptValues` and `ignoreValues` an enum of known options and not a string/regex option. The possible values are known, so we can list them.

Making it an enum makes it easier for others to ignore `block flow-root` and also cover `flow-root block` with the same config. i.e. users won't have to consider the order of keywords in values when configuring this rule.

**Comment by jeddy3** [comment]
I got ahead of myself with all the options. We typically only add them after a user requests them. Let's take out the `*values: []` "user-defined" options for now. We can revisit their behaviour if that use case is ever requested.

We can use an `ignore: ['non-legacy-values']` "keyword" option instead for the first approach in https://github.com/stylelint/stylelint-config-standard/issues/387#issue-3830959229, making it clearer that the initially supported use cases are concerned with legacy values.

I've updated the blueprint above and the first config in https://github.com/stylelint/stylelint-config-standard/issues/387#issue-3830959229.
